File: backends/fastapi/app/api/controllers/auth/register_controller.py
from fastapi import HTTPException, APIRouter, Depends, Response
from app.api.validators.register_validator import RegisterRequest
from app.api.services.user.user_service import UserService
from app.utils.response_formatter import ResponseFormatter
from app.api.services.token.token_service import TokenService

class RegisterController:
    def __init__(self, user_service: UserService, token_service: TokenService):
        self.user_service = user_service
        self.token_service = token_service

    async def register(self, request: RegisterRequest, response: Response = None):
        try:
            user = await self.user_service.register_user(request)
            await self.token_service.generate_verification_token(user.email)
            
            return ResponseFormatter.success("Verification mail sent")

        except HTTPException as e:
            return ResponseFormatter.error(e.detail, e.status_code)
        except Exception as e:
            print(e)
            return ResponseFormatter.error("An unexpected error occurred.", 500)

File: backends/Auther/Controllers/Auth/RegisterController.cs
using Auther.Models;
using Auther.Services.Auth;
using Microsoft.AspNetCore.Mvc;
using Auther.Utilities;

using FluentValidation;

namespace Auther.Controllers.Auth
{
    [ApiController]
    [Route("api/auth/register")]
    public class RegisterController : ControllerBase
    {
        private readonly UserAuthService _userService;
        private readonly IValidator<User> _registerValidator;
        private readonly TokenService _tokenService;

        public RegisterController(UserAuthService userService, TokenService tokenService, IValidator<User> registerValidator)
        {
            _userService = userService;
            _registerValidator = registerValidator;
            _tokenService = tokenService;
        }

        [HttpPost]
        public async Task<IActionResult> Register(User user)
        {
            var validatedFields = await _registerValidator.ValidateAsync(user);

            if (!validatedFields.IsValid)
            {
                return ResponseFormatter.Error(StatusCodes.Status400BadRequest, validatedFields.Errors[0].ErrorMessage);
            }

            try
            {
                user.Role = UserRole.GUEST;
                await _userService.RegisterUserAsync(user);
                string email = user.Email!;
                await _tokenService.GenerateVerificationTokenAsync(email);

                return ResponseFormatter.Success("Verification mail sent");
            }
            catch (ServiceException ex)
            {
                return ResponseFormatter.Error(ex.StatusCode, ex.Message);
            }
            catch (Exception)
            {
                return ResponseFormatter.Error(StatusCodes.Status500InternalServerError, "An unexpected error occurred");
            }
        }
    }
}
